src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/CollectionExtensions.cs

```diff
@@ -57,25 +57,26 @@ public static T[] RemoveLast<T>(this T[] array)
         /// </summary>
         public static ReadOnlyCollection<T> ToReadOnly<T>(this IEnumerable<T>? enumerable)
         {
-            if (enumerable == null)
+            if (enumerable != null && enumerable != ReadOnlyCollection<T>.Empty)
```

Comment:
Nit: couldn't this have been
```suggestion
            if (enumerable is null || enumerable != ReadOnlyCollection<T>.Empty)
            {
                return ReadOnlyCollection<T>.Empty;
            }
```

Reply:
It could, but we also need to return Empty in the case below where the created array has length 0, so doing it this way consolidates those returns rather than having two for Empty.